File: js/src/sorts/quickSort.js
function QuickSort() {

}

QuickSort.prototype = {
  sort: (function() {
    function quickSort(arr) {
      if(arr.length <= 1) {
        return arr;
      }

      var input = arr.slice(0);

      // sorts given array in place
      quickSortHelper(input, 0, input.length - 1);

      return input;
    }

    function quickSortHelper(arr, left, right) {
      var index;

      if(arr.length > 1) {
        index = partition(arr, left, right);

        if(left < index - 1) {
          quickSortHelper(arr, left, index - 1);
        }

        if(index < right) {
          quickSortHelper(arr, index, right);
        }
      }
    }

    function partition(arr, left, right) {
      // select pivot at random
      var pivot = arr[util.getRandomInt(left, right + 1)];

      var i = left;
      var j = right;

      while(i <= j) {
        // find a number left of pivot to swap
        while(arr[i] < pivot) {
          i++;
        }

        // find a number right of pivot to swap
        while(arr[j] > pivot) {
          j--;
        }

        if(i <= j) {
            util.swap(arr, i, j);
            i++;
            j--;
        }
      }

      return i;
    }

    return quickSort;
  }.call(this))
}

File: go/src/sorts/quicksort.go
package sorts

import "math/rand"

func Quicksort(arr []int) []int {
  if len(arr) < 2 { return arr }

  left, right := 0, len(arr) - 1

  pivotIndex := rand.Int() % len(arr)

  // Move the pivot to the right
  arr[pivotIndex], arr[right] = arr[right], arr[pivotIndex]

  // Pile elements smaller than the pivot on the left
  for i := range arr {
    if arr[i] < arr[right] {
      arr[i], arr[left] = arr[left], arr[i]
      left++
    }
  }

  // Place the pivot after the last smaller element
  arr[left], arr[right] = arr[right], arr[left]

  Quicksort(arr[:left])
  Quicksort(arr[left + 1:])

  return arr
}
